mantis-runtime/src/main/java/io/mantisrx/runtime/executor/StageExecutors.java

```diff
@@ -492,31 +479,24 @@ else if (stage instanceof ScalarToGroup) {
             toSink = setupScalarToGroupStage(scalarStage, source, context);
         } else if (stage instanceof KeyToKey) {
             KeyToKey keyToKey = (KeyToKey) stage;
-            Observable<Observable<GroupedObservable<String, T>>> source =
+            Observable<Observable<GroupedObservable<Object, T>>> source =
```

Comment:
I would like thoughts on this. I would ideally like to do this in a better way than just converting to an `Object`.

Reply:
I think key needs to be a string and cannot be arbitrary objects for group stages that are not the last. This may not be enforced anywhere, but stage may fail if key is not a string.

Reply:
I'm actually trying to change that (key limited to a `String` only) with #212 and a follow-up PR to this.
So far my observation is that the limitation is because of the hard-coded key encoders in use. If we move away from that and provide a way for users to specify key encoders, it should be fine.

Reply:
that's fair

Reply:
Some immediate (and half-baked) thoughts that come to mind. They're more intended as discussion starters than suggestions.

1. What about creating a wrapper type `MantisGroupKey<T>` or something? It may still end up being `MantisGroupType<Object>` at this point but at least we can guarantee some interface.
2. Is it possible for us to store the type in the `StageConfig` class and fetch it out?
3. It is possible for us to use reflection here to determine the type?

Reply:
+1 to @codyrioux 's comment above. 
If I understand correctly the 'k1' type info is lost during keyToKey <-> StageConfig input and if it is possible to store and retrieve the k1 type inside KeyToKey (which is a safe cast here with type check already) it would be great. If not then I feel like losing compile-time type check here is not a huge problem as this is something fairly easy to capture in dev phase when applying a wrong key codec.

Reply:
I took another stab at using the correct generic type to preserve the return type. 
Either way, I don't think it matters much because the mantis code doesn't handle generics very well. Moreover, the compiled code loses this type information anyway...

Reply:
I like the idea of moving the key type to `StageConfig` with support for `Void` keytype for scalar stages. #218 does this exact thing by promoting the keyCodec to that stage. I'll take a stab at another refactor in a subsequent PR.